drv/cosmo-hf/src/hf.rs

```diff
@@ -238,6 +247,105 @@ impl ServerImpl {
             }
         }
     }
+
+    // This assumes `begin` and `end` have been bounds checked for overflow
+    // and against the flash clip bounds.
+    fn hash_range_update(
+        &mut self,
+        begin: usize,
+        end: usize,
+    ) -> Result<(), HfError> {
+        let mut buf = [0u8; PAGE_SIZE_BYTES];
+        for addr in (begin..end).step_by(buf.len()) {
+            let size = (end - addr).min(buf.len());
+            // This unwrap is safe because `flash_read` can only fail when given
+            // a lease (where writing into the lease fails if the client goes
+            // away).  Giving it a buffer is infallible.
+            self.drv
+                .flash_read(
+                    // This unwrap is safe because we already checked the range
+                    // when building the initial `begin` address
+                    FlashAddr::new(addr as u32).unwrap_lite(),
+                    &mut &mut buf[..size],
+                )
+                .unwrap_lite();
+            if let Err(e) = self.hash.task.update(size as u32, &buf[..size]) {
+                ringbuf_entry!(Trace::HashUpdateError(e));
+                return Err(HfError::HashError);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn step_hash(&mut self) {
+        match self.hash.state {
+            HashState::Hashing { dev, addr, end } => {
+                let step_size = BLOCK_STEP_SIZE;
+
+                let prev = self.dev;
+                self.set_dev(dev).unwrap();
+                // The only way we should get an error from this is if
+                // we somehow call update before we've initialized or
+                // after we've finished the hash.
+                self.hash_range_update(addr, addr + step_size).unwrap_lite();
+                self.set_dev(prev).unwrap(); // infallible if the earlier set_dev worked
+
+                self.hash.state = if addr + step_size >= end {
+                    HashState::Done
+                } else {
+                    HashState::Hashing {
+                        dev,
+                        addr: addr + step_size,
+                        end,
+                    }
+                };
+
+                match self.hash.state {
+                    HashState::Done => match self.hash.task.finalize_sha256() {
+                        Ok(v) => match dev {
+                            HfDevSelect::Flash0 => {
+                                self.hash.cached_hash0 = SlotHash::Hash(v);
+                            }
+                            HfDevSelect::Flash1 => {
+                                self.hash.cached_hash1 = SlotHash::Hash(v);
+                            }
+                        },
+                        Err(_) => (),
```

Comment:
Log the error in a ringbuf?

Reply:
We don't have a ringbuf on `gimlet-hf` at the moment